// @stripe/ui-extension-sdk/ui/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TableHeaderCell = exports.TableHead = exports.TableFooter = exports.TableCell = exports.TableBody = exports.Tab = exports.TabPanels = exports.TabPanel = exports.TabList = exports.Switch = exports.StripeFileUploader = exports.Spinner = exports.Sparkline = exports.SignInView = exports.SettingsView = exports.Select = exports.Radio = exports.PropertyList = exports.PropertyListItem = exports.PlatformConfigurationView = exports.OnboardingView = exports.Menu = exports.MenuItem = exports.MenuGroup = exports.List = exports.ListItem = exports.Link = exports.LineChart = exports.Inline = exports.Img = exports.Icon = exports.FormFieldGroup = exports.FocusView = exports.Divider = exports.DetailPageTable = exports.DetailPagePropertyList = exports.DetailPageModule = exports.DateField = exports.ContextView = exports.Chip = exports.ChipList = exports.Checkbox = exports.Button = exports.ButtonGroup = exports.Box = exports.BarChart = exports.Banner = exports.Badge = exports.Accordion = exports.AccordionItem = void 0;
exports.Tooltip = exports.TextField = exports.TextArea = exports.TaskList = exports.TaskListItem = exports.Tabs = exports.TableRow = exports.Table = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("@remote-ui/react");
const version_1 = require("../version");
const withSdkProps = (Component) => {
    const wrappedComponentName = Component.displayName || Component.toString();
    const WithSdkProps = (props) => ((0, jsx_runtime_1.jsx)(Component, { ...props, wrappedComponentName: wrappedComponentName, sdkVersion: version_1.SDK_VERSION, schemaVersion: "v9" }));
    WithSdkProps.wrappedComponentName = wrappedComponentName;
    return WithSdkProps;
};
const defineComponent = (name, fragmentProps, wrapWithSdkProps) => {
    const remoteComponent = (0, react_1.createRemoteReactComponent)(name, {
        fragmentProps,
    });
    if (!wrapWithSdkProps) {
        return remoteComponent;
    }
    return withSdkProps(remoteComponent);
};
exports.AccordionItem = defineComponent('AccordionItem', ['title', 'actions', 'media', 'subtitle'], true);
exports.Accordion = defineComponent('Accordion', [], true);
exports.Badge = defineComponent('Badge', [], true);
exports.Banner = defineComponent('Banner', ['actions', 'description', 'title'], true);
exports.BarChart = defineComponent('BarChart', [], true);
exports.Box = defineComponent('Box', [], true);
exports.ButtonGroup = defineComponent('ButtonGroup', ['menuTrigger'], true);
exports.Button = defineComponent('Button', [], true);
exports.Checkbox = defineComponent('Checkbox', ['label'], true);
exports.ChipList = defineComponent('ChipList', [], true);
exports.Chip = defineComponent('Chip', [], true);
exports.ContextView = defineComponent('ContextView', ['actions', 'banner', 'footerContent', 'primaryAction', 'secondaryAction'], true);
exports.DateField = defineComponent('DateField', ['label'], true);
exports.DetailPageModule = defineComponent('DetailPageModule', [], true);
exports.DetailPagePropertyList = defineComponent('DetailPagePropertyList', [], true);
exports.DetailPageTable = defineComponent('DetailPageTable', [], true);
exports.Divider = defineComponent('Divider', [], true);
exports.FocusView = defineComponent('FocusView', ['footerContent', 'primaryAction', 'secondaryAction'], true);
exports.FormFieldGroup = defineComponent('FormFieldGroup', [], true);
exports.Icon = defineComponent('Icon', [], true);
exports.Img = defineComponent('Img', [], true);
exports.Inline = defineComponent('Inline', [], true);
exports.LineChart = defineComponent('LineChart', [], true);
exports.Link = defineComponent('Link', [], true);
exports.ListItem = defineComponent('ListItem', ['icon', 'image', 'secondaryTitle', 'title', 'value'], true);
exports.List = defineComponent('List', [], true);
exports.MenuGroup = defineComponent('MenuGroup', ['title'], true);
exports.MenuItem = defineComponent('MenuItem', [], true);
exports.Menu = defineComponent('Menu', ['trigger'], true);
exports.OnboardingView = defineComponent('OnboardingView', ['error'], true);
exports.PlatformConfigurationView = defineComponent('PlatformConfigurationView', [], true);
exports.PropertyListItem = defineComponent('PropertyListItem', ['label', 'value'], true);
exports.PropertyList = defineComponent('PropertyList', [], true);
exports.Radio = defineComponent('Radio', ['label'], true);
exports.Select = defineComponent('Select', ['label'], true);
exports.SettingsView = defineComponent('SettingsView', [], true);
exports.SignInView = defineComponent('SignInView', ['descriptionActionContents', 'footerContent'], true);
exports.Sparkline = defineComponent('Sparkline', [], true);
exports.Spinner = defineComponent('Spinner', [], true);
exports.StripeFileUploader = defineComponent('StripeFileUploader', [], true);
exports.Switch = defineComponent('Switch', ['label'], true);
exports.TabList = defineComponent('TabList', [], true);
exports.TabPanel = defineComponent('TabPanel', [], true);
exports.TabPanels = defineComponent('TabPanels', [], true);
exports.Tab = defineComponent('Tab', [], true);
exports.TableBody = defineComponent('TableBody', [], true);
exports.TableCell = defineComponent('TableCell', [], true);
exports.TableFooter = defineComponent('TableFooter', [], true);
exports.TableHead = defineComponent('TableHead', [], true);
exports.TableHeaderCell = defineComponent('TableHeaderCell', [], true);
exports.Table = defineComponent('Table', [], true);
exports.TableRow = defineComponent('TableRow', [], true);
exports.Tabs = defineComponent('Tabs', [], true);
exports.TaskListItem = defineComponent('TaskListItem', [], true);
exports.TaskList = defineComponent('TaskList', [], true);
exports.TextArea = defineComponent('TextArea', ['label'], true);
exports.TextField = defineComponent('TextField', ['label'], true);
exports.Tooltip = defineComponent('Tooltip', ['trigger'], true);
